### DISCUSSION TITLE
Target which adds ProjectReference or PackageReference on a condition

### DISCUSSION THREAD

**Discussion by oruchreis** [dotnet/msbuild#11191] [Q&A]
Hi,
I'm trying to add `ProjectReference` or `PackageReference` on a condition with a target. I have csproj files including these:
```xml
<CustomReference Include="Package1" />
<CustomReference Include="Package2" />
```
The condition property is set in `Directory.Build.props`:
```xml
<PropertyGroup>
  <UseLocalNuGetProjects Condition="'$(UseLocalNuGetProjects)' == ''">true</UseLocalNuGetProjects>
```
And the target is implemented in `Directory.Build.targets` :
```xml
<Project>
  <Target Name="ResolveCustomeReferences"  BeforeTargets="BeforeBuild;ResolveReferences">
    <ItemGroup Condition="'@(CustomReference.Identity)' != ''">
      <MSBuild Projects="$(MSBuildThisFileDirectory)Libraries/%(CustomReference.Identity)/%(CustomReference.Identity).csproj" 
               Targets="Restore;Build" Condition="'$(UseLocalNuGetProjects)' == 'true'" />
      <ProjectReference Include="$(MSBuildThisFileDirectory)Libraries/%(CustomReference.Identity)/%(CustomReference.Identity).csproj"
                        Condition="'$(UseLocalNuGetProjects)' == 'true'" IncludeAssets="all" />      
      <PackageReference Include="@(CustomReference)"
                        Condition="'$(UseLocalNuGetProjects)' != 'true'" /><!-- I have also Directory.Packages.props so I don't need version in PackageReference in the target. -->
    </ItemGroup>
  </Target>
</Project>
```

I have 2 problem with this approach:
1. The build is working if the project doesn't use transitive nuget packages in the referenced project with `ProjectReference`. If the project uses nuget packages included in the referenced project with `ProjectReference`, then build fails. Normally, if we set `ProjectReference` statically in the csproj, we can use transitive nuget packages in that reference, but I couldn't overcome the unresolved nuget package errors with above dynamically added `ProjectReference`.
2. The intellisense and solution explorer doesn't see the `ProjectReference` nor `PackageReference` according to `UseLocalNuGetProjects` property.

How can I solve these two problems?

By the way I've tried many targets in `BeforeTargets` like `ResolveProjectReferencesDesignTime2`, `CompileDesignTime`, `ResolveProjectReferences`, `ResolvePackageDependencies`, `Compile`, etc, but no luck.

Best Regards.

**Comment by vst121** [comment]
Hi,

Both issues stem from the fact that ProjectReference and PackageReference items need to be evaluated at project evaluation time, not inside a target. When you add them dynamically inside a target:

- Transitive NuGet references won’t flow correctly because MSBuild resolves NuGet package graphs during evaluation, before any target runs. That’s why your dynamically added ProjectReference doesn’t see the referenced project’s packages.
- IntelliSense / Solution Explorer doesn’t see them because those editors read the project items from the evaluated .csproj, and targets aren’t executed in design-time evaluation.

**Solution:**
- Move your conditional ProjectReference / PackageReference into ItemGroups with conditions directly in your .csproj or 
Directory.Build.props, instead of adding them inside a target. Example:
```
<ItemGroup Condition="'$(UseLocalNuGetProjects)' == 'true'">
  <ProjectReference Include="Libraries\Package1\Package1.csproj" />
  <ProjectReference Include="Libraries\Package2\Package2.csproj" />
</ItemGroup>

<ItemGroup Condition="'$(UseLocalNuGetProjects)' != 'true'">
  <PackageReference Include="Package1" />
  <PackageReference Include="Package2" />
</ItemGroup>
```

- If you need the list to be dynamic, you can use an ItemGroup in Directory.Build.props with <CustomReference> items, then transform them into ProjectReference / PackageReference in another static ItemGroup using MSBuild item metadata and Condition.

This way, NuGet and the IDE can resolve dependencies at evaluation time, and transitive packages will flow correctly.
